pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -166,14 +167,32 @@ func generateMetadataStringMatcher(keys []string, v *metadata.StringMatcher) *me
 	return &metadata.MetadataMatcher{
 		Filter: authn.AuthnFilterName,
 		Path:   paths,
-		Value: &metadata.MetadataMatcher_Value{
-			MatchPattern: &metadata.MetadataMatcher_Value_StringMatch{
+		Value: &metadata.ValueMatcher{
+			MatchPattern: &metadata.ValueMatcher_StringMatch{
 				StringMatch: v,
 			},
 		},
 	}
 }
 
+func generateMetadataListMatcher(keys []string, v *metadata.ListMatcher) *metadata.MetadataMatcher {
+	paths := make([]*metadata.MetadataMatcher_PathSegment, 0)
+	for _, k := range keys {
+		paths = append(paths, &metadata.MetadataMatcher_PathSegment{
+			Segment: &metadata.MetadataMatcher_PathSegment_Key{Key: k},
+		})
+	}
+	return &metadata.MetadataMatcher{
+		Filter: authn.AuthnFilterName,
+		Path:   paths,
+		Value: &metadata.ValueMatcher{
+			MatchPattern: &metadata.ValueMatcher_ListMatch{
+				ListMatch: v,
+			},
+		},
+	}
+}
+
 // createDynamicMetadataMatcher creates a MetadataMatcher for the given key, value pair.
 func createDynamicMetadataMatcher(k, v string) *metadata.MetadataMatcher {
 	var keys []string
```

Comment:
I would suggest to refactor this function. For example, you can break it into 3 parts, the first one generate `key` (would be nice to have a better name, to distinguish with the input key `k`), the second one generate matcher, and third one (this function) to join keys and matchers.

Reply:
Done.